### ISSUE TITLE
Last packets discarded in edge-triggered epoll(again?)

### ISSUE THREAD

**Issue by ninja-** [netty/netty#6550]
### Expected behavior

Hi,
After a little break I am having trouble again with last packets being discarded and it seriously looks like a netty bug(again?) that may be a regression after updating netty after a year of 4.1 updates, but I am unable to tell for sure.

I am using edge-triggered epoll and I am running a custom proxy application for a game
(still the same one that I've been targeting with my netty bugfixes a while ago)

So the schema is
client <-> proxy <-> server

The problem is that when the client is kicked this way:
```
channel.writeAndFlush(kickPacket).addListener(Listeners.CLOSE).;
```

There's a 10% chance the client will not receive the packet and only connection reset.
The only working workaround I was able to create is to call writeAndFlush and call channel.close() 50ms later.....

I seriously have no idea what could be the problem here, but I remember fixing bugs related to this problem a year or two ago.

Another side of the problem is that recently I started seeing this problem on the "server" side so when the server sends a kick packet to the proxy with disconnection reason, the packet is not received before "channelInactive" is called.

The client<->proxy connection is autoread: true, while proxy<->server is on autoread: false to avoid possible backpressure, but both of them are creating a similar problem.
The read() algorithm for the proxy<->server connection is to call channel.read() after last batch of proxied packets have been written via a ChannelPromise
(I am using a dumb promise with empty packet at the end of the flushed batch, so I get a callback after everything has been written)

But even with autoread: false, the logic should be that everything should be read from socket before handling channelActive, unless I am the one calling channel.disconnect(), right? 

Also, the proxy<->client connections are using a watermark like this:
```
                .childOption(ChannelOption.WRITE_BUFFER_HIGH_WATER_MARK, 1024* 1024*5)
                .childOption(ChannelOption.WRITE_BUFFER_LOW_WATER_MARK, 512*1024*2)
```

Any help on where should I start looking for problems(again)?

### Steps to reproduce

I am not sure I am able to produce a good reproducer _yet_ before investigating more.

### Netty version

4.1.9-FINAL-Snapshot as of 4 weeks ago(9ee4cc0ada3d8e46b139671dabfbfd35c8be3308) + write squasher (source below)

### JVM version (e.g. `java -version`)

openjdk version "1.8.0_121"
OpenJDK Runtime Environment (build 1.8.0_121-8u121-b13-0ubuntu1.16.04.2-b13)
OpenJDK 64-Bit Server VM (build 25.121-b13, mixed mode)

**Comment by ninja-** [comment]
Here's a class I am using that I am using to get a "callback" after a batch have been flushed, it might be a source of the bug the class looks legit to me.

```java
/**
 * The idea of a flush marker is to have a callback
 * that is called, after a given batch of packets has been fully written.
 * It's used as a nice way to control the read() function of a proxy when used along with
 * batch packet flushing.
 *
 * When a promise for our "empty packet" is completed
 * that simply means that everything that was before the marker,
 * was written already.
 *
 * A promise needs to be reusable!
 * DO NOT use if you aren't writing your DATA with void promise!
 */
@Getter
@Setter
@RequiredArgsConstructor
class TrafficFlushMarker extends ChannelDuplexHandler {
    private final ChannelPromise promise;
    private boolean something = false;

    @Override
    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
        something = true;
        ctx.write(msg, promise);
    }

    @Override
    public void flush(ChannelHandlerContext ctx) throws Exception {
        if (!something) {
            promise.trySuccess();
            return;
        }

        if (promise != null)
            ctx.write(Unpooled.EMPTY_BUFFER, promise);

        ctx.flush();
        something = false;
    }
}
```

**Comment by ninja-** [comment]
Also I recalled that I am using the following patch to squash the writes but it may be a dead code(at least I wasn't able to get it to work properly so epoll's writeMultiple wouldn't be called 90% of the time, and it still is)

```patch
index beecf78..a83ba9a 100644
--- a/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java
+++ b/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java
@@ -16,6 +16,7 @@
 package io.netty.channel;
 
 import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
 import io.netty.buffer.ByteBufHolder;
 import io.netty.buffer.Unpooled;
 import io.netty.channel.socket.nio.NioSocketChannel;
@@ -89,6 +90,9 @@ public final class ChannelOutboundBuffer {
     private static final AtomicLongFieldUpdater<ChannelOutboundBuffer> TOTAL_PENDING_SIZE_UPDATER =
             AtomicLongFieldUpdater.newUpdater(ChannelOutboundBuffer.class, "totalPendingSize");
 
+    private static final boolean SQUASHER_ENABLED =
+            SystemPropertyUtil.getBoolean("io.netty.voidPromiseWriteCumulator", true); // it's sane to get it enabled by default
+
     @SuppressWarnings("UnusedDeclaration")
     private volatile long totalPendingSize;
 
@@ -109,6 +113,16 @@ public final class ChannelOutboundBuffer {
      * the message was written.
      */
     public void addMessage(Object msg, int size, ChannelPromise promise) {
+        if (SQUASHER_ENABLED && promise.isVoid() && tailEntry != null && tailEntry.promise.isVoid() && msg instanceof ByteBuf && tailEntry.msg instanceof ByteBuf) {
+            ByteBuf input = ((ByteBuf) msg);
+            ByteBuf old = (ByteBuf) tailEntry.msg;
+            if (input.refCnt() == 1) {
+                tailEntry.msg = MERGE_CUMULATOR.cumulate(ByteBufAllocator.DEFAULT, old, input); // our input buffer will be released by the cumulator
+                tailEntry.pendingSize += size;
+                incrementPendingOutboundBytes(size, false);
+                return;
+            }
+        }
         Entry entry = Entry.newInstance(msg, size, total(msg), promise);
         if (tailEntry == null) {
             flushedEntry = null;
@@ -824,4 +838,50 @@ public final class ChannelOutboundBuffer {
             return next;
         }
     }
+
+    // the cumulator copied from ByteToMessageDecoder
+
+    public static final Cumulator MERGE_CUMULATOR = new Cumulator() {
+        @Override
+        public ByteBuf cumulate(ByteBufAllocator alloc, ByteBuf cumulation, ByteBuf in) {
+            ByteBuf buffer;
+            if (cumulation.writerIndex() > cumulation.maxCapacity() - in.readableBytes()
+                    || cumulation.refCnt() > 1) {
+                // Expand cumulation (by replace it) when either there is not more room in the buffer
+                // or if the refCnt is greater then 1 which may happen when the user use slice().retain() or
+                // duplicate().retain().
+                //
+                // See:
+                // - https://github.com/netty/netty/issues/2327
+                // - https://github.com/netty/netty/issues/1764
+                buffer = expandCumulation(alloc, cumulation, in.readableBytes());
+            } else {
+                buffer = cumulation;
+            }
+            buffer.writeBytes(in);
+            in.release();
+            return buffer;
+        }
+    };
+
+    /**
+     * Cumulate {@link ByteBuf}s.
+     */
+    public interface Cumulator {
+        /**
+         * Cumulate the given {@link ByteBuf}s and return the {@link ByteBuf} that holds the cumulated bytes.
+         * The implementation is responsible to correctly handle the life-cycle of the given {@link ByteBuf}s and so
+         * call {@link ByteBuf#release()} if a {@link ByteBuf} is fully consumed.
+         */
+        ByteBuf cumulate(ByteBufAllocator alloc, ByteBuf cumulation, ByteBuf in);
+    }
+
+    static ByteBuf expandCumulation(ByteBufAllocator alloc, ByteBuf cumulation, int readable) {
+        ByteBuf oldCumulation = cumulation;
+        cumulation = alloc.buffer(oldCumulation.readableBytes() + readable);
+        cumulation.writeBytes(oldCumulation);
+        oldCumulation.release();
+        return cumulation;
+    }
+
 }
--
libgit2 0.24.0
```

**Comment by johnou** [comment]
cc @Scottmitch

**Comment by ninja-** [comment]
Also, I have this code in main handler...but I don't think it could cause problems

```java
    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
    {
            if (channel.isClosed())
                return;
         // rest of packet handling here
}
```

**Comment by ninja-** [comment]
ah I remembered why I left the comment about using void promises only...otherwise it gets ignored.
(I needed to check netty's source, because I though the promise would be called after flush and not get attached to the write call)

So that part of the problem should be fixed by removing TrafficFlushMarker before kick, my mistake.

I am still not sure about the part with the server<->proxy kicks not being delivered...

**Comment by ninja-** [comment]
@normanmaurer any ideas this time?

**Comment by normanmaurer** [comment]
@Scottmitch can you please take a look

**Comment by ninja-** [comment]
Duplicate of recently fixed #6303 @Scottmitch? :)

**Comment by johnou** [comment]
@ninja- try 4.1.9.Final, it was released 10th of March.

**Comment by ninja-** [comment]
Yes, it probably fixes the #6303 which is nice. The other half of the problem still remains though :/

**Comment by normanmaurer** [comment]
Please open another bug with only the remaining problem explained
